Add cause to Rejection (WIP)

dmexe opened the pull request:
Discussions: #37, #33 

It doesn't break anything, I've just added an error cause to a Rejection

- [x] Add an error cause to Rejection
- [x] Serialize Rejection into JSON
- [x] Combine rejections with causes
- [x] Use an rejection's error inside filters

seanmonstar commented on src/filters/path.rs:
```diff
@@ -186,15 +186,19 @@ pub fn index() -> impl Filter<Extract=(), Error=Rejection> + Copy {
+pub fn param<T>() -> impl Filter<Extract=One<T>, Error=Rejection> + Copy
+where
+    T: FromStr + Send,
+    T::Err: Into<::reject::Cause>
```
This would be a breaking change, though it definitely seems worth doing as part of 0.2!

seanmonstar commented on src/reject.rs:
```diff
@@ -107,14 +116,47 @@ impl Rejection {
+    }
+}
+
+impl PartialOrd for Rejection {
```
We might not want ordering to be part of the public API of a `Rejection`... Mostly because I don't know if it's obvious to a user *why* it has one. What would they assume the ordering means? The ordering used on the `reason` field can just be an implementation deal in `CombineRejection`.

seanmonstar commented on src/reject.rs:
```diff
@@ -157,23 +203,62 @@ impl Reject for Rejection {
         let mut res = http::Response::default();
         *res.status_mut() = code;
+
+        match serde_json::to_vec(&self) {
```
This may make sense in JSON APIs, but maybe less so in other cases. In other cases, it may be better to just print a plain text message.

The best case would probably be if we could convert this somehow to just `Response<Cause>`, and a solution to #4 would allow a user to automatically add converting rejections into JSON (or some other format)...

seanmonstar commented on src/reject.rs:
```diff
@@ -220,3 +311,75 @@ mod sealed {
+    use super::*;
+
+    #[test]
+    fn combine_rejections() {
```
Thanks so much for adding tests about this! It was sorely missing.

seanmonstar commented on src/filters/body.rs:
```diff
@@ -25,8 +25,9 @@ pub(crate) fn body() -> impl Filter<Extract=(Body,), Error=Rejection> + Copy {
-                Err(reject::server_error())
+                let err = "request body already taken in previous filter";
+                warn!("{}", err);
+                Err(reject::server_error().with(err))
```
This is admittedly pretty cool!

dmexe commented:
Updates summary:
1) created a new `warp::path::param2`, it would be become `warp::param::param` in next major release.
2) removed `PartialCmp` trait implementations, comparison now inside `CombineRejection`
3) replaced an error cause default format from json to text.

seanmonstar commented:
Thanks, this is pretty great!
